### PR TITLE
http: processing of headers with underscores in names

### PR DESCRIPTION
**Pull request by yanavlasov**
* add config option to either allow, drop header or reject request
* applies to client request headers only

Risk Level: Low (default behavior unchanged)
Testing: Unit+Integration Tests
Docs Changes: Yes
Release Notes: Yes

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

File: test/extensions/filters/http/jwt_authn/filter_integration_test.cc [hunk 1]
```diff
@@ -54,7 +54,7 @@ class HeaderToFilterStateFilterConfig : public Common::EmptyHttpFilterConfig {
                                      Server::Configuration::FactoryContext&) override {
     return [](Http::FilterChainFactoryCallbacks& callbacks) -> void {
       callbacks.addStreamDecoderFilter(
-          std::make_shared<HeaderToFilterStateFilter>("jwt_selector", "jwt_selector"));
+          std::make_shared<HeaderToFilterStateFilter>("jwt-selector", "jwt-selector"));
```
Comment (mattklein123) [inline test/extensions/filters/http/jwt_authn/filter_integration_test.cc:57]: Given that we decided to keep the default as allow, maybe revert all of these test changes that we don't need?

**Comment by mattklein123** [comment]
Oops compile error looks legit.

/wait

**Comment by yanavlasov** [comment]
clang-tidy error seems to be bogus. It looks like include search directories may be not set correctly.

https://dev.azure.com/cncf/envoy/_build/results?buildId=36689&view=logs&j=3bf5ec37-bb6f-5fb6-2cf1-5134dc6106d6&t=41d571fe-3811-578d-07de-68468f467dae&l=1570

**Comment by mattklein123** [comment]
@yanavlasov can you make sure that the bazel deps are correct for that target? Sometimes there are issues with that. Also can you merge master and open the follow up issue re: unification and QUIC? Thank you!

/wait

**Comment by yanavlasov** [comment]
Will do. Header validation issue: #10646

**Comment by mattklein123** [comment]
I will go ahead and merge this as we need to do a full master pass over clang-tidy anyway. Thank you!

**Comment by PiotrSikora** [comment]
@mattklein123 @yanavlasov @lizan any thoughts on whether we should backport this to stable branches or not?

**Comment by mattklein123** [comment]
@PiotrSikora I would defer to you. Given that we ultimately agreed with your original assessment that it is a feature request, I think it makes sense either way.

**Pull request merged**